Plug memory leaks, check memory accesses

janisozaur opened the pull request.

duncanspumpkin commented on src/platform/linux.c:
```diff
@@ -592,8 +592,8 @@ void platform_get_user_directory(utf8 *outPath, const utf8 *subDirectory)
 		log_verbose("adding subDirectory '%s'", subDirectory);
-		strcat(buffer, subDirectory);
-		strcat(buffer, separator);
+		strncat(buffer, subDirectory, MAX_PATH);
```
This is just as unsafe as the previous version it can still walk off the end of MAX_PATH. You've corrected this mistake on a strncat lower down.

janisozaur commented:
updated

janisozaur commented:
Oh wait, I just found out there's `strlcat`! Let me use that

janisozaur commented:
Updated to `strlcat`

janisozaur commented:
Using `strlcat` would be problematic and isn't the silver bullet I initially thought it is. I rebased on top of #2386 which changed filenames, fixed few more instances and tested and consider this PR to be in "done" state, so please review & merge.

janisozaur commented:
thanks
